### ISSUE TITLE
Prevent cookie lists from being selected at the same time

### ISSUE THREAD

**Issue by ghost** [uBlockOrigin/uBlock-issues#3231]
### Prerequisites

- [X] I verified that this is not a filter list issue. Report any issues with filter lists or broken website functionality in the [uAssets issue tracker](https://github.com/uBlockOrigin/uAssets/issues).
- [X] This is **NOT** a [_YouTube_](https://www.github.com/uBlockOrigin/uAssets/issues/20586), [_Facebook_](https://www.github.com/uBlockOrigin/uAssets/issues/3367) or [_Twitch_](https://www.github.com/uBlockOrigin/uAssets/issues/5184) report. _These sites MUST be reported by clicking their **respective links**_.
- [X] This is not a support issue or a question. For support, questions, or help, visit [/r/uBlockOrigin](https://www.reddit.com/r/uBlockOrigin/).
- [X] I performed a [cursory search of the issue tracker](https://github.com/uBlockOrigin/uBlock-issues/issues?q=is%3Aissue) to avoid opening a duplicate issue.
- [X] The issue is not present after disabling uBO in the browser.
- [X] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I am reporting is not normal behavior.

### I tried to reproduce the issue when...

- [X] uBO is the only extension.
- [X] uBO uses default lists and settings.
- [X] using a new, unmodified browser profile.

### Description

`AdGuard/uBO – Cookie Notices` and `EasyList/uBO – Cookie Notices` have filters that reload webpages after changing cookie values.

If they're programmed to set cookies to different values, this may cause an infinite refresh loop.

Consider adding a check to prevent them from being selected at the same time. This would prevent conflicts without having to remove either list from the extension. 

If a user has selected `EasyList/uBO – Cookie Notices`, and then click on `AdGuard/uBO – Cookie Notices` , the extension will deselect `EasyList/uBO – Cookie Notices`, and vice-versa.

### A specific URL where the issue occurs.

```yaml
See https://github.com/uBlockOrigin/uAssets/issues/23562
```

### Steps to Reproduce

-

### Expected behavior

-

### Actual behavior

-

### uBO version

1.57.2

### Browser name and version

Firefox 125.0.3

### Operating System and version

Hannah Montana Linux

**Comment by jackrjli** [comment]
This kind of breakage can occur with any list that uses `set-cookie`/`set-cookie-reload`. Making people choose either AdGuard or EasyList Cookie Notices wouldn't solve the issue beyond this particular case, and would get in the way of people with legitimate reasons to use both lists (they don't cover all the same sites).

**Comment by u-RraaLL** [comment]
I wonder if uBO should discard a second filter trying to set the same cookie just as if it were an exact duplicate.

**Comment by stephenhawk8054** [comment]
There were actually other issues prior to the above case with 2 cookie lists conflicted to each other. I agree with the solution of only choosing 1 list for cookie. Other lists can use `set-cookie` but these 2 setlists are the ones using the most extensively and can highly likely cause conflictions, especially when this is the only scriptlet with a `reload` option which can cause more severe issue than other filters.

We don't encourage users to enable more lists to cover different sites any way, and all of the breakages will just be reported to uBO volunteers to solve. Breakage is always a more important issue than lack of site's coverages. Reporting to respective list maintainers is the correct way to solve the latter.

**Comment by u-RraaLL** [comment]
Agreed. In this particular case, though, the user had a custom AG Annoyance added to his lists (instead of enabled AG Cookie Notices itself), so the issue would've still arisen.

**Comment by stephenhawk8054** [comment]
Yes, I think we can just try to cover the most obvious/popular case here since it's impossible to cover all edge cases (even the same list can have same issue if there are some mistakes)

**Comment by gorhill** [comment]
> Consider adding a check to prevent them from being selected at the same time. This would prevent conflicts without having to remove either list from the extension.

This would make it difficult to diagnose filter issues since whichever one takes effect first in undetermined.

**Comment by garry-ut99** [comment]
The most conflicts are between `EasyList` cookies and `AdGuard` cookies, but on the other hand the both lists are highly unique and are great supplements to each other, hence forcing to "enable only one cookie list at once" will significantly reduce the effectiveness of blocking cookie notices, the optimal solution would be to fix conflicts at `set-cookie` scriptlet level if possible.

**Comment by krystian3w** [comment]
In the past set-constant was similar breakage: https://github.com/uBlockOrigin/uBlock-issues/issues/156#issuecomment-709423918

**Comment by gorhill** [comment]
Originally, a long time ago, the AdGuard lists were introduced as stock lists because EasyList lists didn't support extended filtering syntax, limiting itself to only ABP-compatible syntax. It has been a while now that EasyList supports extended syntax, and it might be that AdGuard's lists are no longer needed -- so maybe we should just remove them as stock lists? (Except where there is no alternative)

**Comment by dportvine** [comment]
As far as I can see, the problem will only be with the rules for mobile. Rules for them can be added to EasyList and regional lists. If this is a complex case, add it to uBlock filters.

**Comment by garry-ut99** [comment]
AdGuard filter lists are good supplements for EasyList/uBlock filter lists because they cover big amount of unique websites not covered by EasyList/uBlock filter lists, these websites can be covered in EasyList/uBlock over time, but it will take months or years for EasyList/uBlock filter lists to catch up the similiar coverage. Statistics show that 36% of uBlock users have at least 1 AdGuard filter list enabled.

On the other hand, keeping AdGuard lists in uBlock, causes occasional conflicts/breakages, and inhibits the growth rate of EasyList/uBlock own filters, because less sites are reported to EasyList/uBlock's filter trackers, by users who have AdGuard lists enabled. Removing AdGuard lists from uBlock will cause some of these users to re-add AdGuard lists via "Import" list feature, but some other users will not re-add and will start reporting uncovered sites to EasyList/uBlock filter trackers, helping and assisting in the growth of EasyList/uBlock own filters.

The explanation of why AdGuard dominated EasyList/uBlock filter lists, in number of websites, is:
- AdGuard introduced built-in website reporting feature years ago before uBlock
- AdGuard built-in website reporting feature doesn't require github account to report websites

it caused AdGuard to achieve a massive amount of websites reported: `~170000` issues reported on their [AdGuard filter tracker](https://github.com/AdguardTeam/AdguardFilters/issues), compared to `~41000` on [ EasyList (~21000) + [uAssets](https://github.com/uBlockOrigin/uAssets/issues) (~20000) ] filter trackers combined.

(the calculations for EasyList-related filter reports are the following: 1502 in [Fanboy old tracker](https://github.com/ryanbr/fanboy-adblock/issues)  + [ 7430 in [EasyList new tracker](https://github.com/easylist/easylist/issues) + 22923 [EasyList forum tracker 1](https://forums.lanik.us/viewforum.php?f=62-report-unblocked-content) + 8157 [EasyList forum tracker 2](https://forums.lanik.us/viewforum.php?f=62-report-unblocked-content) ] / 2 = 1502 + 38510 / 2 = `~21000`, it has been divided by 2 because we need to exclude websites reports related to main EasyList, which I don't know how many % it is, so I just blind-shooted to be 50%, because it would be too difficult and time-consuming to calculate precisely).

Now is time to decide: either EasyList/uBlock still includes AdGuard useful lists, or drop them to fix conflicts and to grow faster on its own at the cost of a noticeable reduction in effectiveness (amount of websites covered).

**Comment by Marko-98** [comment]
Just stumbled upon this issue and was thinking about reporting it. Now I see I'm not the only one...

https://github.com/user-attachments/assets/fc26f3f2-1e06-4223-8412-3d02105c92bf

**Comment by krystian3w** [comment]
That is correct, GDPR at this page block load videos: https://github.com/AdguardTeam/AdguardFilters/issues/137106

Therefore, you should check how to unblock the video in EasyList Cookies to eventually finally repair the player in uBo Cookie Notices.

**Comment by Marko-98** [comment]
> That is correct, GDPR at this page block load videos: [AdguardTeam/AdguardFilters#137106](https://github.com/AdguardTeam/AdguardFilters/issues/137106)
> 
> Therefore, you should check how to unblock the video in EasyList Cookies to eventually finally repair the player in uBo Cookie Notices.

Huh, I didn't realize the videos aren't working (I'm not visiting this website often). But I will open an issue on uAssets regarding this.

**Comment by u-RraaLL** [comment]
It'd likely need to compare cookie values against the filters?

**Comment by gorhill** [comment]
> t'd likely need to compare cookie values against the filters?

It's already done: https://github.com/gorhill/uBlock/blob/1.61.3b4/src/js/resources/cookie.js#L109

If reload is required and the current cookie is the same as the target one, the scriptlet bails out.

**Comment by gorhill** [comment]
> Is preventing set-cookie to reload the page more than once in uBO's side feasible?

The scriptlet would need to store something to "remember" it was just executed, and what is stored from page world can be seen by the page, so not a solution.

Since I cannot reproduce the case, I won't be able to analyze what could be done. Signal here if this happens again elsewhere.

**Comment by D4niloMR** [comment]
The same issue happened again: https://www.reddit.com/r/uBlockOrigin/comments/1hbzrji/pixiv_breaking_with_uboagain/ 

AdGuard with `pixiv.net##+js(set-cookie-reload, privacy_policy_agreement, 6)` and EasyList with `pixiv.net##+js(set-cookie, privacy_policy_agreement, 7, , reload, 1)`

**Comment by D4niloMR** [comment]
Yes, then maybe a potential solution is to apply `dontOverwrite` logic by default when using with `set-cookie` with reload parameters.

**Comment by gorhill** [comment]
Declined. At most what can happen is that I drop AdGuard's lists as stock lists. These were included as stock lists in a distant past because EasyList was not making use of uBO's advanced filtering features, which is no longer the case since a long while now.

**Comment by Marko-98** [comment]
@gorhill Why is _uBlock filters – Cookie Notices_ shown twice here? You should definitely move it under _uBlock filters_.

<img width="899" height="348" alt="Image" src="https://github.com/user-attachments/assets/4fddcd4c-419d-40aa-9c2e-fde13daa726e" />

**Comment by u-RraaLL** [comment]
It's supplementary to EL or AG Cookie Notices. It does not function by itself.

**Comment by Marko-98** [comment]
Then it would especially make sense to put it on top of the filter lists under _uBlock filters_.

**Comment by u-RraaLL** [comment]
No, it wouldn't. People wouldn't know to enable it.

uBO's own default lists are grouped together, the rest are in their respective sections. Cookie Notices and Annoyances.

Anyway, if you want to discuss this further, feel free open up a discussion about it. It's not what this issue was about.

**Comment by gorhill** [comment]
> if you want to discuss this further, feel free open up a discussion about it

No point though, changes about this will be declined.
